lib/manageiq/providers/workflows/runner.rb

```diff
@@ -0,0 +1,86 @@
+module ManageIQ
+  module Providers
+    module Workflows
+      class Runner
+        include Vmdb::Logging
+
+        class << self
+          def runner
+            @runner ||= new.tap(&:start)
+          end
+        end
+
+        attr_reader :workflows
+
+        def initialize
+          require "floe"
+          require "concurrent/hash"
+
+          @workflows          = Concurrent::Hash.new
+          @docker_wait_thread = nil
+        end
+
+        def start
+          $workflows_log.debug("Runner: Starting workflows runner...")
+          self.docker_wait_thread = Thread.new { docker_wait }
+          $workflows_log.debug("Runner: Starting workflows runner...Complete")
+        end
+
+        def stop
+          $workflows_log.debug("Runner: Stopping workflows runner...")
+          stop_thread(docker_wait_thread)
+
+          self.docker_wait_thread = nil
+          $workflows_log.debug("Runner: Stopping workflows runner...Complete")
+        end
+
+        def add_workflow(workflow, queue_args)
+          workflows[workflow.id] ||= [workflow, queue_args]
+        end
+
+        def delete_workflow(workflow)
+          workflows.delete(workflow.id)
+        end
+
+        private
+
+        attr_accessor :docker_wait_thread
+
+        def docker_wait
+          loop do
+            docker_runner = Floe::Runner.for_resource("docker")
+            docker_runner.wait do |event, runner_context|
+              $workflows_log.info("Runner: Caught event [#{event}] for container [#{runner_context["container_ref"]}]")
+
+              workflow, queue_args = workflow_by_runner_context(runner_context)
+              next if workflow.nil?
+
+              $workflows_log.info("Runner: Queueing update for WorkflowInstance ID: [#{workflow.id}]")
+
+              workflow.run_queue(**queue_args)
+            end
+          rescue => err
+            $workflows_log.warn("Error: [#{err}]")
+            $workflows_log.log_backtrace(err)
+          end
+        end
+
+        def stop_thread(thread)
+          return if thread.nil?
+
+          thread.kill
+          thread.join(0)
+        end
+
+        def workflow_by_runner_context(runner_context)
+          workflows.detect do |_id, (workflow, _queue_args)|
+            context       = workflow.reload.context
+            container_ref = context.dig("State", "RunnerContext", "container_ref")
```

Comment:
I'm concerned that we only support a single `container_ref`
feeling like we should register by the `container_ref`

But maybe when we need more (map) then we'll just update this.

If we register by container_ref, then we'll need a way to clear out old references (by workflow_id is nice)

Reply:
Yes with Map/Parallel this will have to be tweaked to support that.

Reply:
If we do change this to container_ref, then this method becomes:

```
def workflow_by_runner_context(runner_context)
  workflow.delete(runner_context["container_ref"])
end
```

Reply:
I don't think we can delete from the workflow list on every event, also this assumes we are indexing on the container_ref which I don't think is practical since it can change every state in a workflow